internal/pkg/log/debug.go

```diff
@@ -12,20 +14,97 @@ import (
 // Logs are stored in a buffer by ioutil.Writer.
 type debugLogger struct {
 	*zapLogger
-	*ioutil.Writer
+	all         *ioutil.Writer
+	debug       *ioutil.Writer
+	info        *ioutil.Writer
+	warn        *ioutil.Writer
+	warnOrError *ioutil.Writer
+	error       *ioutil.Writer
+}
+
+// oneLevelEnabler enables only one level. The others are discarded.
+type oneLevelEnabler struct {
+	level zapcore.Level
+}
+
+func (v *oneLevelEnabler) Enabled(level zapcore.Level) bool {
+	return v.level == level
 }
 
 // NewDebugLogger returns logs as string by String() method.
 // See also other methods of the ioutil.Writer.
 func NewDebugLogger() DebugLogger {
-	writer := ioutil.NewBufferedWriter()
-	return &debugLogger{
-		zapLogger: loggerFromZap(zap.New(debugCore(writer))),
-		Writer:    writer,
+	l := &debugLogger{
+		all:         ioutil.NewBufferedWriter(),
+		debug:       ioutil.NewBufferedWriter(),
+		info:        ioutil.NewBufferedWriter(),
+		warn:        ioutil.NewBufferedWriter(),
+		warnOrError: ioutil.NewBufferedWriter(),
+		error:       ioutil.NewBufferedWriter(),
+	}
+	l.zapLogger = loggerFromZap(zap.New(zapcore.NewTee(
+		debugCore(l.all, DebugLevel),                            // all = debug level and higher
+		debugCore(l.debug, &oneLevelEnabler{level: DebugLevel}), // only debug msgs
+		debugCore(l.info, &oneLevelEnabler{level: InfoLevel}),   // only info msgs
+		debugCore(l.warn, &oneLevelEnabler{level: WarnLevel}),   // only warn msgs
+		debugCore(l.warnOrError, WarnLevel),                     // warn or error = warn level and higher
+		debugCore(l.error, ErrorLevel),                          // error = error level and higher
+	)))
+	return l
+}
+
+// ConnectTo connects all messages to a writer, for example os.Stdout.
+func (l *debugLogger) ConnectTo(writer io.Writer) {
+	l.all.ConnectTo(writer)
+}
+
+// Truncate clear all messages.
+func (l *debugLogger) Truncate() {
+	for _, w := range l.allWriters() {
+		w.Truncate()
 	}
 }
 
-func debugCore(writer *ioutil.Writer) zapcore.Core {
+// AllMsgs returns all messages and Truncate all messages.
+func (l *debugLogger) AllMsgs() string {
```

Comment:
Mno, já teda zkratkám v názvech proměnných a metod moc nefandím. 🙂  Ono je to sice doporučení z nějakýho phpkovýho standardu, ale imho to platí obecně. Zhoršuje to rychlost pochopení o co jde, je to náchylnější na překlepy a tak. Ale samozřejmě to závisí i na individuálních preferencích, tak nenutím jen navrhuju. 😃

Reply:
Idealne by sa ta metoda volala `All/Info/...`, ... ale ta je uz pouzita na logovanie (zapis), takze potrebujeme iny nazov.

V Go sa preferuju `Short but Descriptive` nazvy, hlavne sa to tyka variables.
https://betterprogramming.pub/naming-conventions-in-go-short-but-descriptive-1fa7c6d2f32a

Teda napr. jednopismenkove premenne su OK, ak su v limitovanom kontexte, napr. jedna `for loop`, alebo par riadkov kodu.

Teraz je otazka ci `msgs` je dostatocne `descriptive`, alebo ako ju nazvat lepsie.
Ked som to pisal, tak mi to prislo OK, ... ale ked to tak tebe nepride, tak to bude lepsie upravit.

Podla mna je dolezitejsie to `descriptive`, ako to ci je to skratka alebo nie.
Napr. `Fs()` mi pride ovela lepsie ako `Filesystem()`, hlavne ak sa to pouziva casto a riadky by boli zbytocne dlhsie.
A to zase znamena dlhsie precitanie a pochopenie kodu.

V kode to potom vyzera takto:
```go
assert.Empty(t, logger.WarnAndErrorMsgs())
```

Po zmene
```go
assert.Empty(t, logger.WarnAndErrorMessages())
```

Dlzka nazvu metody `WarnAndErrorMessages` pri pride uz dost dlha, ale este OK, tak to upravim.

Vdaka za koment ;)

Reply:
https://github.com/keboola/keboola-as-code/pull/418/commits/eb7626d3eed382aa071e419f921fc73e83f5aefe